Inconsistency between &HashMap and &BTreeMap implementation of IntoIterator

The implementation of `IntoIterator` for `&'a BTreeMap<K, V>` and `&'a HashMap<K, V, S>` differ in the lifetime bounds of `K` and `V`

```rust
impl<'a, K: 'a, V: 'a> IntoIterator for &'a BTreeMap<K, V> { \* ... *\ }
impl<'a, K, V, S> IntoIterator for &'a HashMap<K, V, S> { \* ... *\ }
```

I believe those additional bounds should be unnecessary since they're implied by the fact that exists a reference of lifetime `'a` to the whole map. However, quite interestingly, they appear to change how the compiler resolves trait bounds in some cases. For example in the following code:

```rust
trait Map
where
    for<'a> &'a Self: IntoIterator<Item = (&'a Self::Key, &'a Self::Value)>,
{
    type Key;
    type Value;
}

// Ok, compiles
impl<K, V> Map for HashMap<K, V> {
    type Key = K;
    type Value = V;
}

// Not ok, doesn't compiles, with rather cryptic error
impl<K, V> Map for BTreeMap<K, V> {
    type Key = K;
    type Value = V;
}
```

I would expect it to compile, or at least throw the same error for both the `BTreeMap` and `HashMap` implementation. 
However it throws errors only for the `BTreeMap` implementation. 

Also note that error E0311 isn't even present in the Rust Compiler Error Index.

```
error[E0311]: the parameter type `K` may not live long enough
  --> src/lib.rs:18:12
   |
18 | impl<K, V> Map for BTreeMap<K, V> {
   |            ^^^
   |
   = help: consider adding an explicit lifetime bound for `K`
   = note: the parameter type `K` must be valid for any other region...
note: ...so that the type `K` will meet its required lifetime bounds
  --> src/lib.rs:18:12
   |
18 | impl<K, V> Map for BTreeMap<K, V> {
   |            ^^^

error[E0311]: the parameter type `V` may not live long enough
  --> src/lib.rs:18:12
   |
18 | impl<K, V> Map for BTreeMap<K, V> {
   |            ^^^
   |
   = help: consider adding an explicit lifetime bound for `V`
   = note: the parameter type `V` must be valid for any other region...
note: ...so that the type `V` will meet its required lifetime bounds
  --> src/lib.rs:18:12
   |
18 | impl<K, V> Map for BTreeMap<K, V> {
   |            ^^^

error: aborting due to 2 previous errors
```

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c214377f7f3d85b8f193aef4015e89fe)

Found a more minimal test case:

```rust
fn foo<'a, K, V, I: 'a + IntoIterator<Item = (&'a K, &'a V)>>(i: I) {}

fn bar() {
    let map = HashMap::<u32, u32>::new();
    foo(&map); // Ok
    
    let map = BTreeMap::<u32, u32>::new();
    foo(&map); // Error
}
```

The error it gives is pretty much the same:

```
error[E0309]: the parameter type `K` may not live long enough
 --> src/lib.rs:3:39
  |
3 | fn foo<'a, K, V, I: 'a + IntoIterator<Item = (&'a K, &'a V)>>(i: I) {}
  |            -                          ^^^^^^^^^^^^^^^^^^^^^
  |            |
  |            help: consider adding an explicit lifetime bound...: `K: 'a`
  |
note: ...so that the reference type `&'a K` does not outlive the data it points at
 --> src/lib.rs:3:39
  |
3 | fn foo<'a, K, V, I: 'a + IntoIterator<Item = (&'a K, &'a V)>>(i: I) {}
  |                                       ^^^^^^^^^^^^^^^^^^^^^

error[E0309]: the parameter type `V` may not live long enough
 --> src/lib.rs:3:39
  |
3 | fn foo<'a, K, V, I: 'a + IntoIterator<Item = (&'a K, &'a V)>>(i: I) {}
  |               -                       ^^^^^^^^^^^^^^^^^^^^^
  |               |
  |               help: consider adding an explicit lifetime bound...: `V: 'a`
  |
note: ...so that the reference type `&'a V` does not outlive the data it points at
 --> src/lib.rs:3:39
  |
3 | fn foo<'a, K, V, I: 'a + IntoIterator<Item = (&'a K, &'a V)>>(i: I) {}
  |                                       ^^^^^^^^^^^^^^^^^^^^^
```

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5f24ea404ba2933cb718384230e7fcd6)

--- a/library/alloc/src/collections/btree/map.rs
+++ b/library/alloc/src/collections/btree/map.rs
@@ -1294,7 +1294,7 @@ impl<K: Ord, V> BTreeMap<K, V> {
 }
 
 #[stable(feature = "rust1", since = "1.0.0")]
-impl<'a, K: 'a, V: 'a> IntoIterator for &'a BTreeMap<K, V> {
+impl<'a, K, V> IntoIterator for &'a BTreeMap<K, V> {
     type Item = (&'a K, &'a V);
     type IntoIter = Iter<'a, K, V>;
 
@@ -1363,7 +1363,7 @@ impl<K, V> Clone for Iter<'_, K, V> {
 }
 
 #[stable(feature = "rust1", since = "1.0.0")]
-impl<'a, K: 'a, V: 'a> IntoIterator for &'a mut BTreeMap<K, V> {
+impl<'a, K, V> IntoIterator for &'a mut BTreeMap<K, V> {
     type Item = (&'a K, &'a mut V);
     type IntoIter = IterMut<'a, K, V>;
 

--- a/src/test/ui/btreemap/btreemap_into_iterator_lifetime.rs
+++ b/src/test/ui/btreemap/btreemap_into_iterator_lifetime.rs
@@ -1,0 +1,23 @@
+// check-pass
+
+use std::collections::{BTreeMap, HashMap};
+
+trait Map
+where
+    for<'a> &'a Self: IntoIterator<Item = (&'a Self::Key, &'a Self::Value)>,
+{
+    type Key;
+    type Value;
+}
+
+impl<K, V> Map for HashMap<K, V> {
+    type Key = K;
+    type Value = V;
+}
+
+impl<K, V> Map for BTreeMap<K, V> {
+  type Key = K;
+  type Value = V;
+}
+
+fn main() {}

